### ISSUE TITLE
proposal: move slave syncer out of tlogserver

### ISSUE THREAD

**Issue by GlenDC** [zero-os/0-Disk#596]
// TODO:
// add intro, description, problems with current approach and finally solutions (propositions)

**Comment by iwanbk** [comment]
Very interesting proposal, some quick into will be very helpful especially about the solution.
As slave syncer is indeed not normally a task for a tlog server.

If we can see some raw solution proposal soon, we might make this issue and #587 to block each other.

**Comment by GlenDC** [comment]
Sure thing. Will add it tomorrow as soon as possible. Today I was too busy with coding work. Sorry for that.

**Comment by GlenDC** [comment]
@iwanbk I added the description. It's quite a lot to take in though, and it's a bit chaotic as it's written directly from my mind. However I hope it helps us to think and decide about the future of the slave sync feature.

Feel free to discuss it tomorrow with others such as @zaibon as well.

**Comment by iwanbk** [comment]
> However I think we all agreed since the beginning that syncing (meta)data to the slave cluster shouldn't be the responsibility of the tlogserver. That is also something highlighted by @robvanmieghem's comment from back in May: #38 (comment). There was however not a bad idea back than, and so we continued with the idea of implementing this slave syncer component within the tlogserver, using the tlogplayer.

Yep, it is correct.
Please also note that slave syncer also make the tlogserver slower

> There is no more extra communication (start/stop slave sync) required between nbdserver and tlogserver

For now, the extra communication from nbdserver is only wait for syncing process to be finished

> The slave syncer becomes a lot cheaper as there is no longer an entire tlogreplay process (which has to decode, decrypt, decompress, etc...)

No
The entire tlogreplay process only happens when the data source is from 0-stor.
if the data comes from tlogclient directly, tlogreplay don't need to decrypt & decompress.

**Comment by iwanbk** [comment]
> An alternative solution that i had in mind would be to have some kind of message queue service (similar to or RabbitMQ itself). 

Instead of rabbitmq, we might want something more lightweight like [nsq](nsq.io)

> This would be used to the nbdserver for all data write calls. Than it would write all that data to the primary cluster, slave cluster and tlogserver.

Sounds reasonable for slave and tlog, but not sure about writing to primary cluster from an MQ.

**Comment by GlenDC** [comment]
> Please also note that slave syncer also make the tlogserver slower

Interesting to note indeed. It makes the argument of going for the tlogserver-backed solution because of performance a lot weaker. As it could be argued that the performance cost has just been moved from the nbdserver to the tlogserver. While the total cost is still at least as much, if not more.

> For now, the extra communication from nbdserver is only wait for syncing process to be finished

Sure but that doesn't matter. In the final design (if we should continue with the slave syncer as a component within the tlogserver) we require much more complex communication, as we need to be able to disable slave syncing to specific servers of a cluster when the nbdserver takes control over them, and only enable those servers for slave syncing again when the nbdserver no longer desires such direct control.

> Instead of rabbitmq, we might want something more lightweight like nsq

Sure, or even a very simplistic custom solution. I was just saying RabbitMQ because of the concept of message queueing and how it executes that concept. Didn't really want to point to any implementation in specific, it was more about the idea.

> Sounds reasonable for slave and tlog, but not sure about writing to primary cluster from an MQ.

Yeah it's going to cause problems. As there will quite a big window when the data is not yet in the actual ARDB, while it has already left the storage. But those same issues are also problematic for the slave servers, as we might switch at any point to those servers so that window is just as bad for those. And I would argue that the same also goes for the tlog data.

I wouldn't focus to much on the alternative MQ solution if I were you @iwanbk. I think it's more of a random side thought, that can probably be ignored.

> The entire tlogreplay process only happens when the data source is from 0-stor.

That's interesting. Even though it doesn't make me any more of a fan of the current slave sync implementation. Still for the sake of completeness, could you point me to the current master code where you are writing "the data [which] comes from tlogclient directly", as you say. As I seem to recall that the only slave cluster writing happens using the tlogplayer.

**Comment by iwanbk** [comment]
> Interesting to note indeed. It makes the argument of going for the tlogserver-backed solution because of performance a lot weake

Yeah, because more task for the CPU

> "the data [which] comes from tlogclient directly"

just to make sure, it means from tlogclient-> tlogserver -> slave syncer

>  As I seem to recall that the only slave cluster writing happens using the tlogplayer.

Well, those decompress etc is done by 0-stor client, only needed when tlogplayer got data from 0-stor.
tlogplayer has API to replay data directly from an aggregation https://godoc.org/github.com/zero-os/0-Disk/tlog/tlogclient/player#Player.ReplayAggregationWithCallback


data is being sent to slave syncer here
https://github.com/zero-os/0-Disk/blob/master/tlog/tlogserver/server/vdisk.go#L470

received here 
https://github.com/zero-os/0-Disk/blob/master/tlog/tlogserver/slavesync/slave_sync.go#L254

and then replayed with the func i listed above

**Comment by GlenDC** [comment]
Yeah I know more or less how _that_ code looks, I've seen it plenty of times. But I have yet to see slave-sync code that writes data directly to the slave cluster as it is received from the tlogclient in the tlogserver.

**Comment by GlenDC** [comment]
Also we're getting too much in detail here. And it all doesn't matter too much as I think we all agree that going for the current solution (having the slave syncer as a component of the tlogserver) isn't a solution we want to roll with.

So instead of going back and forward about details like this, I would rather focus on finding and agreeing on a new slave syncer implementation, which is completely decoupled from the tlog module. In the spirit of this I would encourage you to think about this and propose solutions as you come up with them. Also like I said, it might be useful to discus them with others of the team while I'm offline for the night.

**Comment by iwanbk** [comment]
i edited the issue description, crossing the point about "need entire tlogreplay" and add point about "slavesync make tlogserver to be slower"

**Comment by GlenDC** [comment]
@iwanbk what do you in fact think about my actual proposed solution (where the slave-syncer is just an async part of the normal nbdserver vdisk storage code). Because I don't think you've shared your opinion on that one yet?

Also did you have any (alternation of) ideas of your own about the future of the slave syncer?

**Comment by iwanbk** [comment]
> @iwanbk what do you in fact think about my actual proposed solution (where the slave-syncer is just an async part of the normal nbdserver vdisk storage code)

Basically i'm agree with this 

> Because I don't think you've shared your opinion on that one yet?

Yeah, because this proposal still very vague, hard to comment at this stage.

> Also did you have any (alternation of) ideas of your own about the future of the slave syncer?

going to think about it today.

> With the design as it currently is however, layer 2 will fail as soon as there is a problem in layer 3. Which makes it that layer 2 and layer 3 are in fact one and the same in the current design.

Actually, write to layer 1 will also failed if layer 3 failed.

**Comment by iwanbk** [comment]
After some thinking today, i think we could implement my (improved) first idea, proposed at https://github.com/zero-os/0-Disk/issues/38#issuecomment-302987755.
In which tlogserver don't fill the slave, but only assist the slave

So, we can setup slave cluster using redis/ardb replication mechanism, and let the master do the syncing to the slave.
In case of master become failed:
- check value of nbd's last flushed sequence(lfs) in slave, the value was added as part of #230 
- execute tlogreplay to write data from sequence lfs+1 to the end.

The good part of this idea are:
- should be faster than addding slave syncer to tlog or nbd
- simpler of course
- could be implemented as `zeroctl` command that could be invoked by the orchestrator
 
@robvanmieghem criticized the original version of this idea at https://github.com/zero-os/0-Disk/issues/38#issuecomment-303419968

> @iwanbk @GlenDC Was thinking about that too originally. Thing is you need to know more or less to what point the slave caught up, monitor the master slave replication, detect a failing master, trigger this process, put a new config next to the nbdserver and SIGHUP it, all without causing io errors in the vm using the nbdserver.

Except the first two point, all other points remain true (well, except the SIGHUP part, but `etcd` part is basically the same).

For the first two point : 
- thing is you need to know more or less to what point the slave caught up : with nbd's last flushed sequence metadata, we know the point
- monitor the master slave replication : we don't need to monitor it

For the other points:
we have to do it anyway, even with current slave syncer

**Comment by GlenDC** [comment]
> Actually, write to layer 1 will also failed if layer 3 failed.

Yeah currently. But that shouldn't be happening imho. Hence what I was talking about in https://github.com/zero-os/0-Disk/pull/521#issuecomment-343496064 (I was by the way not proposing to do that work as part of that PR, it simply seemed the most natural place to bring up those thoughts).

**Comment by GlenDC** [comment]
I think the problem with using ardb/redis master/slave replication is that we have no control over it at all. Can we at any point know how far the slave has caught up with master? I suppose starting/stopping master/slave replication is possible, but as nbdserver doesn't know anything about the ARDB server processes, it cannot do so on its own and will have to do that via AYS, which is an issue as that is a delayed process.

In short, I really don't see that working with our current infrastructure.

**Comment by iwanbk** [comment]
> I think the problem with using ardb/redis master/slave replication is that we have no control over it at all. 

what is `we` here? AYS or nbdserver?
If ays, i guess they have

> Can we at any point know how far the slave has caught up with master?

Yeah, it is on my proposal
```
thing is you need to know more or less to what point the slave caught up : with nbd's last flushed sequence metadata, we know the point
```

>  I suppose starting/stopping master/slave replication is possible, but as nbdserver doesn't know anything about the ARDB server processes, it cannot do so on its own and will have to do that via AYS, which is an issue as that is a delayed process.

How long is the delay?

**Comment by GlenDC** [comment]
> what is we here? AYS or nbdserver? If ays, i guess they have

AYS does, but it doesn't just _instantly_ react on stuff that propagate from nbdserver down you know. And neither is there currently a way for us to wait on some kind of answer from AYS. In fact there is not even a way for AYS to communicate back to us, other than to update our configs or shut is down.

So even if AYS has control over ardb/redis master/slave replication (which indeed it probably does have), there is no way that the nbdserver could suddenly turn on/off replication for a specific server. We could request it via an STDERR log to AYS, but than there would be a certain window where both the NBDServer could be writing to a slave server and the slave server itself is still being replicated from the primary server.

Well thinking about this, I guess this is perhaps not an issue, as we only switch to writing to slave server directly in case the primary server is down. And if that's the case than I guess the replication process can't be writing to the slave server anymore either, as there is no data to be replicated, as the primary server is down.

> Yeah, it is on my proposal

In your proposal we would still have a highly coupled layer 2 (slave cluster) and layer 3 (tlog), which is entirely what we're trying to prevent with this issue... I think...

> How long is the delay?

Who knows, as if you can put a number on that.

**Comment by iwanbk** [comment]
> In your proposal we would still have a highly coupled layer 2 (slave cluster) and layer 3 (tlog), which is entirely what we're trying to prevent with this issue... I think...

Hmm.. not sure with your response.

What i mean is https://github.com/zero-os/0-Disk/blob/master/nbd/ardb/storage/tlog.go#L37

If it is also what you mean, well, while it is only used by `tlogStorage` this is not really specific to tlog, because it is last flushed to ardb not to tlog.

If what you mean is we still need to use tlogreplay, then i was thinking yesterday to use a kind of buffer in nbdserver. The buffer entry will be erased once it knows that it already synced to slave.

**Comment by iwanbk** [comment]
> How long is the delay?
> Who knows, as if you can put a number on that.

I mean, we also have delay with any async solution

**Comment by zaibon** [comment]
My thought about this topic:

Regarding the solutoin propsed by @GlenDC to move slave syncer to nbdserver:

I mostly agree with this idea cause it cut out all the hard communication that needs to happens between tlogserver  and nbdserver. Which I think is can be a source of a lot problem and edge cases.
Now I have few question:

> It might hurt the performance of the nbdserver, as it now has to write data to 2 clusters per vdisk (which supports the slave cluster), instead of just 1;

How would it have to impact performance, the write to the slave cluster has to be asynchronous.

> It makes it harder to guarantee that the slave cluster is in sync with the primary clusters;

Why is that ? we still know what has been flushed or not with this design no ? To me this design also simplify that since the information stays local to the nbdserver

> If an nbdserver crashes it would potentially make both the primary- and slave cluster corrupt, 

This is not something you can prevent I think. If nbdserver crash, then the last valid state is the one from the last transaction that is on the tlog. This is the design

> Introducion new queue is indeed something I wouldn't do. this bring complexity and that's what we're trying to avoid by keeping syncer in the nbdserver


Now some question regarding some comment from @iwanbk :
> Please also note that slave syncer also make the tlogserver slower

Why is that? This should be done asynchronous, so no delay for the tlog server.

nbdserver shouldn't not wait for the tlog to have replicated the data on the slave cluster. But it should be able to know what it needs to still replay in case nbdserver needs to swtich to slave cluster. so tlog server can tell nbdserver, ok now you can start using slave cluster, it has the same state as your master before the crash.
- Please also note that slave syncer also make the tlogserver slower
Why is that? this should be done asynchronous, so no delay for the tlog server

- nbdserver shouldn't not wait for the tlog to have replicated the data on the slave cluster. But it should be able to know what it needs to still replay in case nbdserver needs to swtich to slave cluster. so tlog server can tell nbdserver, "ok now you can start using slave cluster, it has the same state as your master before the crash" But like I said this kind of communication seems pretty complex and error prone.

**Comment by iwanbk** [comment]
> How would it have to impact performance, the write to the slave cluster has to be asynchronous.

> Why is that? This should be done asynchronous, so no delay for the tlog server.

Async doesn't mean zero performance hit. It still takes your CPU and RAM.
But yeah, less hit than the sync one.

**Comment by GlenDC** [comment]
> Why is that ? we still know what has been flushed or not with this design no ? 

Yeah I'm talking more about in case we restart an nbdserver. Or mount an existing vdisk. There would currently be no way to know that the slave cluster is up to date with the primary cluster.

I do like the idea to implement the slave cluster syncing as a subscriber to the tlog stream, but at the same time that would mean that we would only provide the slave cluster support for vdisks that make use of the tlog repo. Than again I suppose that doesn't have to be a problem, as it all falls under the umbrella of data-redudancy, it might be fine if this is like this. Even so, we would still need to somehow have a way to validate (at startup/failover) that a slave server/cluster is up to date with the primary counterpart.
